Opus and Vorbis webm aren't available in the content process on some platforms.

jyavenard opened the pull request:
#### 5e174d0a75e215fe0987f4966d6ccb1502075ced
<pre>
Opus and Vorbis webm aren&apos;t available in the content process on some platforms.
<a href="https://bugs.webkit.org/show_bug.cgi?id=300013">https://bugs.webkit.org/show_bug.cgi?id=300013</a>
<a href="https://rdar.apple.com/161800673">rdar://161800673</a>

Reviewed by Youenn Fablet.

We check the Opus and Vorbis capabilities on creation of the GPU process.
This value is then cached in the UI process and passed on follow-up WebContent
process creation; re-using the infrastucture that used to exist for VP9 and AV1
hardware detection.

We simplify the code and combine all the media codecs information into a single
GPUProcessMediaCodecCapabilities structure.

Added test for opus.
* LayoutTests/media/media-can-play-webm-expected.txt:
* LayoutTests/media/media-can-play-webm.html:
* Source/WebCore/platform/graphics/cocoa/WebMAudioUtilitiesCocoa.h:
* Source/WebCore/platform/graphics/cocoa/WebMAudioUtilitiesCocoa.mm:
(WebCore::setHasOpusDecoder):
(WebCore::setHasVorbisDecoder):
(WebCore::isOpusDecoderAvailable):
(WebCore::isVorbisDecoderAvailable):
* Source/WebKit/CMakeLists.txt:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.cpp:
(WebKit::m_sharedPreferencesForWebProcess):
* Source/WebKit/Shared/GPUProcessConnectionParameters.h:
* Source/WebKit/Shared/GPUProcessConnectionParameters.serialization.in:
* Source/WebKit/Shared/GPUProcessMediaCodecCapabilities.h: Copied from Source/WebKit/Shared/GPUProcessConnectionParameters.h.
* Source/WebKit/Shared/GPUProcessMediaCodecCapabilities.serialization.in: Copied from Source/WebKit/WebProcess/GPU/GPUProcessConnectionInfo.serialization.in.
* Source/WebKit/UIProcess/GPU/GPUProcessProxy.cpp:
(WebKit::GPUProcessProxy::createGPUProcessConnection):
(WebKit::GPUProcessProxy::didFinishLaunching):
* Source/WebKit/UIProcess/GPU/GPUProcessProxy.h:
* Source/WebKit/UIProcess/GPU/GPUProcessProxy.messages.in:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/GPU/GPUProcessConnection.cpp:
(WebKit::GPUProcessConnection::didInitialize):
* Source/WebKit/WebProcess/GPU/GPUProcessConnectionInfo.h:
* Source/WebKit/WebProcess/GPU/GPUProcessConnectionInfo.serialization.in:

Canonical link: <a href="https://commits.webkit.org/300939@main">https://commits.webkit.org/300939@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ed0f18acfb35643dd68216384cdd35ec0aa26c18

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/124400 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/44080 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/34810 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/131239 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/76427 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/aff112c5-0f35-4d5f-a526-1e48ff930bb6) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/126277 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/44826 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/52679 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/94636 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/76427 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/0fa930a4-3a1f-4374-9b04-ba2d20bc07b6) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/127354 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/35729 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/111264 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/75218 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/574a9b09-b821-4fb3-a419-eb014193974f) 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/34669 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/29426 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/74716 "Built successfully") | | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/105490 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/29646 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/133904 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/51298 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/39134 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/103122 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [❌ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/51697 "Failed to checkout and rebase branch from PR 51673") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/107482 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/102905 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26191 "Built successfully and passed tests") | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/26522 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/48236 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/51165 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/56956 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/50590 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/53954 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/52262 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebKit/GPUProcess/GPUConnectionToWebProcess.cpp:
```diff
@@ -357,35 +361,34 @@ GPUConnectionToWebProcess::GPUConnectionToWebProcess(GPUProcess& gpuProcess, Web
-        hasVP9HardwareDecoder = WebCore::vp9HardwareDecoderAvailableInProcess();
-        gpuProcess.send(Messages::GPUProcessProxy::SetHasVP9HardwareDecoder(hasVP9HardwareDecoder));
-    }
+        capabilities = GPUProcessMediaCodecCapabilities {
```
```suggestion
        capabilities = {
```

youennf replied:
I would have written it as:
```
auto capabilities = parameters.mediaCodecCapabilities.value_or({
});
```

jyavenard committed:
Opus and Vorbis webm aren't available in the content process on some platforms.
https://bugs.webkit.org/show_bug.cgi?id=300013
rdar://161800673

Reviewed by Youenn Fablet.

We check the Opus and Vorbis capabilities on creation of the GPU process.
This value is then cached in the UI process and passed on follow-up WebContent
process creation; re-using the infrastucture that used to exist for VP9 and AV1
hardware detection.

We simplify the code and combine all the media codecs information into a single
GPUProcessMediaCodecCapabilities structure.

Added test for opus.
* LayoutTests/media/media-can-play-webm-expected.txt:
* LayoutTests/media/media-can-play-webm.html:
* Source/WebCore/platform/graphics/cocoa/WebMAudioUtilitiesCocoa.h:
* Source/WebCore/platform/graphics/cocoa/WebMAudioUtilitiesCocoa.mm:
(WebCore::setHasOpusDecoder):
(WebCore::setHasVorbisDecoder):
(WebCore::isOpusDecoderAvailable):
(WebCore::isVorbisDecoderAvailable):
* Source/WebKit/CMakeLists.txt:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.cpp:
(WebKit::m_sharedPreferencesForWebProcess):
* Source/WebKit/Shared/GPUProcessConnectionParameters.h:
* Source/WebKit/Shared/GPUProcessConnectionParameters.serialization.in:
* Source/WebKit/Shared/GPUProcessMediaCodecCapabilities.h: Copied from Source/WebKit/Shared/GPUProcessConnectionParameters.h.
* Source/WebKit/Shared/GPUProcessMediaCodecCapabilities.serialization.in: Copied from Source/WebKit/WebProcess/GPU/GPUProcessConnectionInfo.serialization.in.
* Source/WebKit/UIProcess/GPU/GPUProcessProxy.cpp:
(WebKit::GPUProcessProxy::createGPUProcessConnection):
(WebKit::GPUProcessProxy::didFinishLaunching):
* Source/WebKit/UIProcess/GPU/GPUProcessProxy.h:
* Source/WebKit/UIProcess/GPU/GPUProcessProxy.messages.in:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/GPU/GPUProcessConnection.cpp:
(WebKit::GPUProcessConnection::didInitialize):
* Source/WebKit/WebProcess/GPU/GPUProcessConnectionInfo.h:
* Source/WebKit/WebProcess/GPU/GPUProcessConnectionInfo.serialization.in:

Canonical link: https://commits.webkit.org/300939@main
